extension-adinfo/src/adinfo.cpp

```diff
@@ -6,27 +6,42 @@
 
 #define DLIB_LOG_DOMAIN LIB_NAME
 #include <dmsdk/sdk.h>
-#include "private_adinfo.h"
+#include "adinfo.h"
+
+
+
+
+ADInfoData g_ADInfoData;
+
+void ADInfo_QueueAdId(const char* adId) {
+    g_ADInfoData.m_AdId = adId;
+}
 
 static int GetAdInfo(lua_State* L) {
     DM_LUA_STACK_CHECK(L, 1);
-    const char* id = ADInfo_GetAdId();
-    if (id) 
+
+    if(g_ADInfoData.m_callback != 0)
     {
-        lua_newtable(L);
-        lua_pushliteral(L, "ad_ident");
-        lua_pushstring(L, id);
-        lua_rawset(L, -3);
-        lua_pushliteral(L, "ad_tracking_enabled");
-        lua_pushboolean(L, ADInfo_IsAdvertisingTrackingEnabled());
-        lua_rawset(L, -3);
+        dmLogError("Callback already in progress");
+        lua_pushboolean(L, 0);
+        return 1;
     }
-    else
+
+    luaL_checktype(L, 1, LUA_TFUNCTION);
+    if (g_ADInfoData.m_callback) {
+        dmScript::DestroyCallback(g_ADInfoData.m_callback);
+    }
```

Comment:
Will this ever occur, given the if statement on line 23?

Reply:
Good point :-)